teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+
+### Use Cases
+
+This TEP contributes to the achievements of the following use cases:
+
+- Achieve SLSA L3 supply chain security through Tekton `PipelineRuns`
+- End-to-end trust of the provenance produced by Tekton Chains for Tekton `PipelineRuns`
+- `Tasks` to emit a list of artifacts of unknown length.
+
+### Requirements
+
+- Enable `Tasks` to be able to surface Artifact metadata as an array.
+
+## Proposal
+
+Define the structure of the Artifact metadata that the `Step` needs to write out. Define the file path and variable replacement syntax for the location of the metadata file.
+
+The following chapters in the proposal describe in details the proposed phased implementation approach:
+
+- [Artifact Provenance Data](#artifact-metadata)
+- [surfacing artifact metadata into the status](#from-step-into-status)
+- [surfacing the artifact metadata into the provenance payload](#from-status-into-provenance)
+
+
+### Artifact Provenance Data
+
+An artifact can be consumed or produced by a `Task`. This means that we will always have two categories at a high level: `inputs` (to indicate what was downloaded) and `outputs` (to indicate what was uploaded). The metadata content could be defined as follows:
+
+```json
+{
+    "inputs":[
+        "name": "artifact Name", # e.g. source code, dependencies
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ],
+    "outputs": [
+        "name": "artifact Name", # e.g. images, packages, coverage reports
+        "values": [
+            {
+                "uri": string,
+                "digest": string
+            },
+        ]
+    ]
+}
+```
+The metadata content is written by the `Step` to a file `$(step.artifacts.path)`:
+
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee $(step.artifacts.path)
+```
+The path `$(step.artifacts.path)` resolves to `/tekton/steps/<step-name>/artifacts/provenance.json` in the `Step` container at runtime.
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee /tekton/steps/artifact-reporting/artifacts/provenance.json
+```
+
+#### Name
```

Comment:
Can you add some examples of how this would work? 

This is a bit different than how we're been thinking about this before, where tasks could output artifact data in the background. And having a name then means we need to define what happens if multiple things try to write in the same namespace (particularly across steps)

Reply:
I think StepActions changes the picture a bit. Here, I'm imagining that a single Step will provide the metadata information about some artifact. Therefore, the artifact will be a step-level thing instead of a Task level. And as such even if multiple steps have artifacts with name conflicts, we can still distinguish them because of different steps. Within the same step, they cannot have the same name.

Reply:
I made the `name` field optional like we discussed in the Data Interfaces WG. PTAL.